torchdata/datapipes/iter/util/randomsplitter.py

```diff
@@ -0,0 +1,171 @@
+        target: Optional key (that must exist in ``weights``) to indicate the specific group to return.
+            If set to the default ``None``, returns ``List[IterDataPipe]``.
+            If target is specified, returns ``IterDataPipe``.
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> dp = IterableWrapper(range(10))
+        >>> train, valid = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0)
+        >>> list(train)
+        [2, 3, 5, 7, 8]
+        >>> list(valid)
+        [0, 1, 4, 6, 9]
+        >>> # You can also specify a target key if you only need a specific group of samples
+        >>> train = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0, target='train')
+        >>> list(train)
+        [2, 3, 5, 7, 8]
+        >>> # Be careful to use the same seed as before when specifying `target` to get the correct split.
+        >>> valid = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0, target='valid')
+        >>> list(valid)
+        [0, 1, 4, 6, 9]
+    """
+
+    def __new__(
+        cls,
+        source_datapipe: IterDataPipe,
+        weights: Dict[T, Union[int, float]],
+        seed,
+        total_length: Optional[int] = None,
+        target: Optional[T] = None,
+    ):
+        if total_length is None:
+            try:
+                # TODO: This is an issue for DataPipes which only have runtime lengths. Revisit to see if this
+                #       is problematic.
+                total_length = len(source_datapipe)
+            except TypeError:
+                raise TypeError(
+                    "RandomSplitter needs `total_length`, but it is unable to infer it from "
+                    f"the `source_datapipe`: {source_datapipe}."
+                )
+        container = _RandomSplitterIterDataPipe(source_datapipe, total_length, weights, seed)
+        if target is None:
+            return [SplitterIterator(container, k) for k in list(weights.keys())]
+        else:
+            if target in weights.keys():
+                return SplitterIterator(container, target)
+            else:
+                raise KeyError(f"`target={target}` does not match any key in `weights`.")
+
+
+class _RandomSplitterIterDataPipe(IterDataPipe):
+    def __init__(
+        self,
+        source_datapipe: IterDataPipe,
+        total_length: int,
+        weights: Dict[T, Union[int, float]],
+        seed,
+    ):
+        self.source_datapipe: IterDataPipe = source_datapipe
+        self.total_length = total_length
+        self._seed = seed
+        self.norm_weights = self.normalize_weights(weights, total_length)
+        self.keys = list(weights.keys())
+        self.key_to_index = {k: i for i, k in enumerate(self.keys)}
+        self.weights = [self.norm_weights[k] for k in self.keys]
```

Comment:
nit: I feel those variables can be simplified.
Why don't we just get a copy of `self.norm_weights`? `key_to_index` seems redundant to me

Reply:
I ended up doing that because [`random.choices`](https://docs.python.org/3/library/random.html#random.choices) (used in `draw`) take in two `list` as arguments.

It is more efficient to convert `self.keys` and `self.weights` to `list` once here rather than for every draw:
`selected_key = self._rng.choices(list(self.weights.keys()), list(self.weights.values()))[0]`

Reply:
I see. Then, `self.norm_weights` can also be a list since we can rely on `keys` and `key_to_index` to indicate the offset. Super nit. Feel free to ignore.